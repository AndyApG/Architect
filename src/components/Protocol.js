import React, { useEffect } from 'react';
import { connect, useDispatch } from 'react-redux';
import { motion } from 'framer-motion';
import cx from 'classnames';
import PropTypes from 'prop-types';
import { getHasUnsavedChanges } from '@selectors/session';
import { selectors as statusSelectors } from '@modules/ui/status';
import { actionCreators as dialogActions } from '@modules/dialogs';
import { actionLocks as protocolsLocks } from '@modules/userActions';
import { actionCreators as sessionActions } from '@modules/session';
import Overview from '@components/Overview';
import Timeline from '@components/Timeline';
import ProtocolControlBar from '@components/ProtocolControlBar';

const Protocol = ({
  isLoading,
  hasProtocol,
  protocolPath,
}) => {
  const dispatch = useDispatch();
  useEffect(() =>
    () => dispatch(sessionActions.resetSession()),
  [protocolPath]);

  const sceneClasses = cx(
    'scene',
    { 'scene--protocol': hasProtocol },
    { 'scene--loading': isLoading },
  );

  const variants = {
    show: {
      opacity: 1,
      transition: {
        duration: 0.5,
      },
    },
    hide: {
      opacity: 0,
    },
  };

  return (
    <motion.div
      className={sceneClasses}
      variants={variants}
    >
      <div className="scene__protocol">
        <Overview
          show={hasProtocol}
        />
        <Timeline show={hasProtocol} />
      </div>
      <ProtocolControlBar show={hasProtocol} />
    </motion.div>
  );
};

Protocol.propTypes = {
  isLoading: PropTypes.bool,
  hasProtocol: PropTypes.bool,
  protocolPath: PropTypes.string,
};

Protocol.defaultProps = {
  protocolPath: null,
  isLoading: false,
  hasProtocol: false,
};

const mapStateToProps = (state) => {
  const activeProtocol = state.session.filePath;

  return {
    hasUnsavedChanges: getHasUnsavedChanges(state),
    protocolPath: activeProtocol,
    hasProtocol: !!activeProtocol,
    isLoading: statusSelectors.getIsBusy(state, protocolsLocks.loading),
  };
};

const mapDispatchToProps = {
  openDialog: dialogActions.openDialog,
};

const withStore = connect(mapStateToProps, mapDispatchToProps);

export { Protocol };

export default withStore(Protocol);
