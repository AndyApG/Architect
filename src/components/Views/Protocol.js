import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
// import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import cx from 'classnames';
import { pick, get } from 'lodash';
import history from '../../history';
import { Button, Icon } from '../../ui/components';
import { getProtocol } from '../../selectors/protocol';
import ShowRoute from '../../components/ShowRoute';
import EditSkipLogic from '../Cards/EditSkipLogic';
import EditStage from '../Cards/EditStage';
import EditForm from '../Cards/EditForm';
import VariableRegistry from '../Cards/VariableRegistry';
import ViewForms from '../Cards/ViewForms';
import EditType from '../Cards/EditType';
import Timeline from '../../components/Timeline';
import ControlBar from '../ControlBar';
// import { actionCreators as protocolFileActions } from '../../ducks/modules/protocol/file';

const RightArrow = <Icon name="arrow-right" />;

class Protocol extends PureComponent {
  static propTypes = {
    overview: PropTypes.shape({
      title: PropTypes.string,
      version: PropTypes.string,
      forms: PropTypes.object,
      variableRegistry: PropTypes.object,
    }).isRequired,
    stages: PropTypes.array.isRequired,
    hasUnsavedChanges: PropTypes.bool,
    hasChanges: PropTypes.bool,
    match: PropTypes.object,
    location: PropTypes.object.isRequired,
    saveProtocol: PropTypes.func.isRequired,
    isProtocol: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    hasUnsavedChanges: false,
    hasChanges: false,
    match: null,
  };

  constructor(props) {
    super(props);

    this.state = {
      new: {
        insertAtIndex: null,
        type: null,
      },
    };
  }

  get protocolPath() {
    return get(this.props.match, 'params.protocol');
  }

  handleRouteComplete = (goto) => {
    if (this.protocolPath) {
      switch (goto) {
        case 'protocol':
          history.push(`/edit/${this.protocolPath}/`);
          break;
        default:
          history.goBack();
      }
    }
  }

  goto = (location = '') => {
    history.push(`/edit/${this.protocolPath}/${location}`);
  }

  createStage = (type, insertAtIndex) =>
    this.goto(`stage?type=${type}&insertAtIndex=${insertAtIndex}`);

  editStage = stageId =>
    this.goto(`stage/${stageId}`);

  editSkipLogic = stageId =>
    this.goto(`skip/${stageId}`);

  render() {
    const {
      overview,
      stages,
      location,
      hasChanges,
      hasUnsavedChanges,
      saveProtocol,
      isProtocol,
    } = this.props;

    if (!isProtocol) { return null; }

    const protocolClasses = cx(
      'protocol',
      {
        'protocol--has-changes': hasChanges,
        'protocol--has-unsaved-changes': hasUnsavedChanges,
      },
    );
    return (
      <div className={protocolClasses}>
        <Timeline
          overview={overview}
          stages={stages}
          location={location}
          onEditStage={this.editStage}
          onEditSkipLogic={this.editSkipLogic}
          onCreateStage={this.createStage}
          hasUnsavedChanges={hasUnsavedChanges}
        />

        <ControlBar show={hasUnsavedChanges}>
          <Button
            size="small"
            onClick={saveProtocol}
            color="white"
            icon={RightArrow}
            iconPosition="right"
          >
            Save
          </Button>
        </ControlBar>

        <ShowRoute
          path={'/edit/:protocol/stage/:id?'}
          location={location}
          component={EditStage}
          onComplete={this.handleRouteComplete}
        />

        <ShowRoute
          path={'/edit/:protocol/skip/:id'}
          location={location}
          component={EditSkipLogic}
          onComplete={this.handleRouteComplete}
        />

        <ShowRoute
          path={'/edit/:protocol/form(s?)'}
          location={location}
          component={ViewForms}
          onComplete={() => this.handleRouteComplete('protocol')}
        />

        <ShowRoute
          path={'/edit/:protocol/form/:form?'}
          location={location}
          component={EditForm}
          onComplete={this.handleRouteComplete}
        />

        <ShowRoute
          path="/edit/:protocol/registry"
          location={location}
          component={VariableRegistry}
          onComplete={() => this.handleRouteComplete('protocol')}
        />

        <ShowRoute
          path="/edit/:protocol/registry/:category/:type?"
          location={location}
          component={EditType}
          onComplete={this.handleRouteComplete}
        />
      </div>
    );
  }
}

function mapStateToProps(state, props) {
  const protocol = getProtocol(state);
  const overview = pick(protocol, ['name', 'version', 'variableRegistry', 'externalData', 'forms']);

  return {
    isProtocol: !!props.match,
    overview,
    stages: protocol.stages,
    hasChanges: state.protocol.past.length > 0,
    hasUnsavedChanges: (state.session.lastChanged > state.session.lastSaved),
  };
}

function mapDispatchToProps() {
  return {
    saveProtocol: () => {},
    // saveProtocol: bindActionCreators(protocolFileActions.saveProtocol, dispatch),
  };
}

export { Protocol };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Protocol);
