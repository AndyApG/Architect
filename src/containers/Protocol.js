/* eslint-disable */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import cx from 'classnames';
import { Button, Icon } from '../ui/components';
import { getProtocol } from '../selectors/protocol';
import EditSkipLogic from '../containers/EditSkipLogic';
import EditStage from '../containers/EditStage';
import { Timeline } from '../components';
import { actionCreators as protocolFileActions } from '../ducks/modules/protocol/file';
import { actionCreators as protocolsActions } from '../ducks/modules/protocols';

const cards = {
  newStage: Symbol('newStage'),
  editSkip: Symbol('editSkip'),
};

const defaultActiveCardState = {
  cardType: null,
  cancel: false,
};

const RightArrow = <Icon name="arrow-right" />;

class Protocol extends PureComponent {
  static propTypes = {
    stages: PropTypes.array.isRequired,
    hasUnsavedChanges: PropTypes.bool,
    hasChanges: PropTypes.bool,
    match: PropTypes.object.isRequired,
    loadProtocol: PropTypes.func.isRequired,
    saveProtocol: PropTypes.func.isRequired,
    exportProtocol: PropTypes.func.isRequired,
  };

  static defaultProps = {
    hasUnsavedChanges: false,
    hasChanges: false,
  };

  constructor(props) {
    super(props);

    this.state = {
      activeCard: { ...defaultActiveCardState },
    };
  }

  componentDidMount() {
    const protocolPath = decodeURIComponent(this.props.match.params.protocol);
    this.props.loadProtocol(protocolPath);
  }

  onCardComplete = () => {
    this.setState({
      activeCard: {
        ...defaultActiveCardState,
      },
    });
  }

  onCardCancel = () => {
    this.setState({
      activeCard: {
        ...defaultActiveCardState,
        cancel: true,
      },
    });
  }

  showCard = (card, { ...options }) => {
    this.setState({
      activeCard: {
        ...defaultActiveCardState,
        cardType: card,
        ...options,
      },
    });
  }

  editStage = stageId => this.showCard(cards.editStage, { stageId });
  createStage = (type, insertAtIndex) => this.showCard(cards.editStage, { type, insertAtIndex });

  isAnyCardVisible = () => this.state.activeCard.cardType !== null;
  isCardVisible = cardType => this.state.activeCard.cardType === cardType;
  isTimelineVisible = () => !this.isAnyCardVisible();

  render() {
    const {
      stages,
      hasChanges,
      hasUnsavedChanges,
      saveProtocol,
      exportProtocol,
    } = this.props;

    const protocolClasses = cx(
      'protocol',
      {
        'protocol--has-changes': hasChanges,
        'protocol--has-unsaved-changes': hasUnsavedChanges,
      },
    );
    return (
      <div className={protocolClasses}>
        <Timeline
          stages={stages}
          onEditSkipLogic={stageId => this.showCard(cards.editSkip, { stageId })}
          onEditStage={this.editStage}
          onCreateStage={this.createStage}
          hasUnsavedChanges={hasUnsavedChanges}
        />

        <div className="protocol__control-bar">
          <Button
            size="small"
            onClick={exportProtocol}
            color="white"
            icon={RightArrow}
            iconPosition="right"
          >
            Export
          </Button>
          { hasUnsavedChanges &&
            <Button
              size="small"
              onClick={saveProtocol}
              color="white"
              icon={RightArrow}
              iconPosition="right"
            >
              Save
            </Button>
          }
        </div>

        <EditSkipLogic
          show={this.isCardVisible(cards.editSkip)}
          cancel={this.state.activeCard.cancel}
          stageId={this.state.activeCard.stageId}
          onComplete={this.onCardComplete}
          onCancel={this.onCardCancel}
        />

        <EditStage
          {...this.state.activeCard} // either index & type, or id
          show={this.isCardVisible(cards.editStage)}
          onComplete={this.onCardComplete}
          onCancel={this.onCardCancel}
        />
      </div>
    );
  }
}

function mapStateToProps(state) {
  const protocol = getProtocol(state);

  return {
    stages: protocol.stages,
    hasChanges: state.protocol.past.length > 0,
    hasUnsavedChanges: (state.session.lastChanged > state.session.lastSaved),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    // saveProtocol: bindActionCreators(protocolFileActions.saveProtocol, dispatch),
    // exportProtocol: bindActionCreators(protocolFileActions.exportProtocol, dispatch),
    loadProtocol: bindActionCreators(protocolsActions.loadProtocol, dispatch),
  };
}

export { Protocol };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Protocol);
