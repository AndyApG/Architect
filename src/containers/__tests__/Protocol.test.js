/* eslint-env jest */

import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import { Protocol } from '../Protocol';

let subject = null;

const mockProps = {
  stages: [{ id: 1, type: 'Foo' }],
};

const mockStore = () =>
  createStore(
    () => ({
      protocol: { present: { options: { title: '' } } },
    }),
  );

const makeSubject = () =>
  mount(
    <Provider store={mockStore()}>
      <Protocol {...mockProps} store={mockStore()} />
    </Provider>,
  );

describe('<Protocol />', () => {
  it('can render', () => {
    subject = shallow(<Protocol {...mockProps} />);

    expect(subject).toMatchSnapshot();
  });

  it('intial state', () => {
    subject = shallow(<Protocol {...mockProps} />);
    subject.update();
    expect(subject.find('NewStage').length).toEqual(0);
  });

  describe('Screen change interactions', () => {
    beforeEach(() => {
      subject = makeSubject();
    });

    it('Shows/hides add stage screen', () => {
      subject.find('TimelineAddNew').find('button').first().simulate('click');
      expect(subject.find('NewStage').prop('show')).toBe(true);
      subject.find('NewStage').find('button').last().simulate('click');
      expect(subject.find('NewStage').prop('cancel')).toBe(true);
      expect(subject.find('NewStage').prop('show')).toBe(false);
    });

    it('Shows/hides edit skip logic screen');

    it('Hides add stage screen when screen added');

    it('Hides edit skip logic screen when saving');
  });
});
