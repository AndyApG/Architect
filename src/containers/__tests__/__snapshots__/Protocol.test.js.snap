// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Protocol /> can render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Protocol
    exportProtocol={[Function]}
    hasChanges={false}
    hasUnsavedChanges={false}
    loadProtocol={[Function]}
    match={
        Object {
            "params": Object {},
          }
    }
    saveProtocol={[Function]}
    stages={
        Array [
            Object {
              "id": 1,
              "type": "Foo",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withContext(Timeline)
          hasUnsavedChanges={false}
          onCreateStage={[Function]}
          onEditSkipLogic={[Function]}
          onEditStage={[Function]}
          stages={
                    Array [
                              Object {
                                "id": 1,
                                "type": "Foo",
                              },
                            ]
          }
/>,
        <div
          className="protocol__control-bar"
/>,
        <Connect(Draft(EditSkipLogic))
          cancel={false}
          onCancel={[Function]}
          onComplete={[Function]}
          show={false}
          stageId={undefined}
/>,
        <Connect(EditStage)
          cancel={false}
          cardType={null}
          onCancel={[Function]}
          onComplete={[Function]}
          show={false}
/>,
      ],
      "className": "protocol",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "hasUnsavedChanges": false,
          "onCreateStage": [Function],
          "onEditSkipLogic": [Function],
          "onEditStage": [Function],
          "stages": Array [
            Object {
              "id": 1,
              "type": "Foo",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "protocol__control-bar",
        },
        "ref": null,
        "rendered": false,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cancel": false,
          "onCancel": [Function],
          "onComplete": [Function],
          "show": false,
          "stageId": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cancel": false,
          "cardType": null,
          "onCancel": [Function],
          "onComplete": [Function],
          "show": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withContext(Timeline)
            hasUnsavedChanges={false}
            onCreateStage={[Function]}
            onEditSkipLogic={[Function]}
            onEditStage={[Function]}
            stages={
                        Array [
                                    Object {
                                      "id": 1,
                                      "type": "Foo",
                                    },
                                  ]
            }
/>,
          <div
            className="protocol__control-bar"
/>,
          <Connect(Draft(EditSkipLogic))
            cancel={false}
            onCancel={[Function]}
            onComplete={[Function]}
            show={false}
            stageId={undefined}
/>,
          <Connect(EditStage)
            cancel={false}
            cardType={null}
            onCancel={[Function]}
            onComplete={[Function]}
            show={false}
/>,
        ],
        "className": "protocol",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "hasUnsavedChanges": false,
            "onCreateStage": [Function],
            "onEditSkipLogic": [Function],
            "onEditStage": [Function],
            "stages": Array [
              Object {
                "id": 1,
                "type": "Foo",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "protocol__control-bar",
          },
          "ref": null,
          "rendered": false,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cancel": false,
            "onCancel": [Function],
            "onComplete": [Function],
            "show": false,
            "stageId": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cancel": false,
            "cardType": null,
            "onCancel": [Function],
            "onComplete": [Function],
            "show": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
